File: MonteCarloTreeSearch/src/main/java/game/board/Board.java
package game.board;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import game.game.Move;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
public class Board {
	private Piece[][] boardDisplay;
	
	public Board(Piece[][] boardDisplay) {
		super();
		this.boardDisplay = boardDisplay;
	}
	

	public static Board of(Piece[][] boardDisplay) {
		return new Board(boardDisplay);
	}

	public static Board of(Integer[][] integerDisplay){
		Piece[][] boardDisplay = new Piece[integerDisplay.length][integerDisplay[0].length];
		for(int i = 0; i<integerDisplay.length; i++){
			Integer[] integerRow = integerDisplay[i];
			for(int j = 0; j<integerRow.length; j++){
				switch (integerRow[j]) {
					case 0:
						boardDisplay[i][j] = Piece.EMPTY;
						break;
					case 1:
						boardDisplay[i][j] = Piece.BLACK_PAWN;
						break;
					case 2:
						boardDisplay[i][j] = Piece.WHITE_PAWN;
						break;
					case 3:
						boardDisplay[i][j] = Piece.WHITE_KING;
						break;
				
					default:
						break;
				}
			}
		}
		return Board.of(boardDisplay);

	}


	@Override
	public Board clone() {
		return new Board(this.boardDisplay);
	}

	public Integer getNumOfRows(){
		return boardDisplay.length;
	}

	public Integer getNumOfColumns(){
		return boardDisplay[0].length;
	}

	public Coordinate getThrone(){
		return new Coordinate((getNumOfRows()-1)/2, (getNumOfColumns()-1)/2);
	}

	public boolean isSurroundingsOfThrone(Coordinate c){
		return c.equals(new Coordinate(((getNumOfRows()-1)/2)-1, (getNumOfColumns()-1)/2)) ||
				c.equals(new Coordinate(((getNumOfRows()-1)/2)+1, (getNumOfColumns()-1)/2)) ||
				c.equals(new Coordinate((getNumOfRows()-1)/2, ((getNumOfColumns()-1)/2)-1)) ||
				c.equals(new Coordinate((getNumOfRows()-1)/2, ((getNumOfColumns()-1)/2)+1));
	}





	 /**
	* 
	* @param variant The variant of the game is going to be played
	* @return the initial state of the board with said variant
	*/
	public static Board getInitialStateOfVariant(Integer variant){
		Integer[][] integerDisplay =  
		{{0,0,1,1,1,0,0}, 
		{0,0,0,1,0,0,0}, 
		{1,0,2,2,2,0,1}, 
		{1,1,2,3,2,1,1}, 
		{1,0,2,2,2,0,1},
		{0,0,0,1,0,0,0},
		{0,0,1,1,1,0,0}};

		return Board.of(integerDisplay);
	}




	/**
	* 
	* @return the coordinates in control of each player
	*/
	public Map<PlayerColor, List<Coordinate>> getPlayerAvailablePieces(){
		Map<PlayerColor, List<Coordinate>> res = new HashMap<PlayerColor,List<Coordinate>>();

		for(int i = 0; i<boardDisplay.length; i++){
			Piece[] row = boardDisplay[i];
			for(int j = 0; j<row.length; j++){
				Piece piece = row[j];
				if(piece.equals(Piece.BLACK_PAWN)){
					if(res.containsKey(PlayerColor.BLACK)){
						List<Coordinate> coordinateList = res.get(PlayerColor.BLACK);
						coordinateList.add(new Coordinate(i, j));
						res.put(PlayerColor.BLACK, coordinateList);
					}else{
						List<Coordinate> coordinateList = new ArrayList<Coordinate>();
						coordinateList.add(new Coordinate(i, j));
						res.put(PlayerColor.BLACK, coordinateList);
					}
				}else if(piece.equals(Piece.WHITE_KING) || piece.equals(Piece.WHITE_PAWN)){
					if(res.containsKey(PlayerColor.WHITE)){
						List<Coordinate> coordinateList = res.get(PlayerColor.WHITE);
						coordinateList.add(new Coordinate(i, j));
						res.put(PlayerColor.WHITE, coordinateList);
					}else{
						List<Coordinate> coordinateList = new ArrayList<Coordinate>();
						coordinateList.add(new Coordinate(i, j));
						res.put(PlayerColor.WHITE, coordinateList);
					}
				}

			}
		}

		return res;
	}


	/**
	* 
	* @return the moves available for each player in a position
	*/
	public Map<PlayerColor, Set<Move>> getPlayerAvailableMoves(){

		Map<PlayerColor, Set<Move>> res = new HashMap<PlayerColor,Set<Move>>();
		Integer numOfRows = this.getNumOfRows();
		Integer numOfColumns = this.getNumOfColumns();

		for(int i = 0; i<numOfRows; i++){
			for(int j = 0; j<numOfColumns; j++){
				Piece piece = boardDisplay[i][j];
				Set<Move> moves = Board.getAvailableMovesOfPiece(new Coordinate(i, j), boardDisplay);
				if(piece.equals(Piece.BLACK_PAWN)){
					if(res.containsKey(PlayerColor.BLACK)){
						Set<Move> moveList = res.get(PlayerColor.BLACK);
						moveList.addAll(moves);
						res.put(PlayerColor.BLACK, moveList);
					}else{
						res.put(PlayerColor.BLACK, moves);
					}
				}else if(piece.equals(Piece.WHITE_KING) || piece.equals(Piece.WHITE_PAWN)){
					if(res.containsKey(PlayerColor.WHITE)){
						Set<Move> moveList = res.get(PlayerColor.WHITE);
						moveList.addAll(moves);
						res.put(PlayerColor.WHITE, moveList);
					}else{
						res.put(PlayerColor.WHITE, moves);
					}
				}
			}
		}
		return res;
	}



	/**
	* @param currentCoordinate The coordinate of the piece that we want to check
	* @param boardDisplay The distribution of pieces in the board
	* @return the moves available for the piece contained in said coordinate
	*/
	private static Set<Move> getAvailableMovesOfPiece(Coordinate currentCoordinate, Piece[][] boardDisplay) {

		Set<Move> res = new HashSet<Move>();
		Integer numOfRows = boardDisplay.length;
		Integer numOfColumns = boardDisplay[0].length;
		Piece piece = boardDisplay[currentCoordinate.getRowIndex()][currentCoordinate.getColumIndex()];

		List<Coordinate> keySquares = new ArrayList<Coordinate>();
		keySquares.add(new Coordinate(0, 0));
		keySquares.add(new Coordinate(numOfRows-1, 0));
		keySquares.add(new Coordinate(0, numOfColumns-1));
		keySquares.add(new Coordinate(numOfRows-1, numOfColumns-1));
		keySquares.add(new Coordinate((numOfRows-1)/2, (numOfColumns-1)/2));


		if(piece.equals(Piece.EMPTY))
			return res;

		res.addAll(checkLeftMoves(piece, currentCoordinate, numOfColumns, numOfRows, boardDisplay, keySquares));
		res.addAll(checkRightMoves(piece, currentCoordinate, numOfColumns, numOfRows, boardDisplay, keySquares));
		res.addAll(checkUpMoves(piece, currentCoordinate, numOfColumns, numOfRows, boardDisplay, keySquares));
		res.addAll(checkDownMoves(piece, currentCoordinate, numOfColumns, numOfRows, boardDisplay, keySquares));
		
		return res;
	}


	private static Collection<? extends Move> checkDownMoves(Piece piece, Coordinate currentCoordinate, Integer numOfColumns, Integer numOfRows, Piece[][] boardDisplay, List<Coordinate> keySquares) {
		List<Move> res = new ArrayList<Move>();
		for(int i = currentCoordinate.getRowIndex()+1; i<numOfRows; i++){
			Move move = Move.of(currentCoordinate.getRowIndex(), currentCoordinate.getColumIndex(), i, currentCoordinate.getColumIndex());
			if(!boardDisplay[i][currentCoordinate.getColumIndex()].equals(Piece.EMPTY))
				break;
			if(keySquares.contains(new Coordinate(i, currentCoordinate.getColumIndex())) && !piece.equals(Piece.WHITE_KING))
				continue;
				if(currentCoordinate.getColumIndex() == 0 || currentCoordinate.getColumIndex()==numOfColumns-1){
					res.add(move);
				}else{
					if(isBlack(piece) && (!isWhite(boardDisplay[i][currentCoordinate.getColumIndex()+1]) || !isWhite(boardDisplay[i][currentCoordinate.getColumIndex()-1])))
						res.add(move);
					if(isWhite(piece) && (!isBlack(boardDisplay[i][currentCoordinate.getColumIndex()+1]) || !isBlack(boardDisplay[i][currentCoordinate.getColumIndex()-1])))
						res.add(move);
				}
		}
		return res;
	}


	private static Collection<? extends Move> checkUpMoves(Piece piece, Coordinate currentCoordinate, Integer numOfColumns, Integer numOfRows, Piece[][] boardDisplay, List<Coordinate> keySquares) {
		List<Move> res = new ArrayList<Move>();
		for(int i = currentCoordinate.getRowIndex()-1; i>=0; i--){
			Move move = Move.of(currentCoordinate.getRowIndex(), currentCoordinate.getColumIndex(), i, currentCoordinate.getColumIndex());
			if(!boardDisplay[i][currentCoordinate.getColumIndex()].equals(Piece.EMPTY))
				break;
			if(keySquares.contains(new Coordinate(i, currentCoordinate.getColumIndex())) && !piece.equals(Piece.WHITE_KING))
				continue;
			if(currentCoordinate.getColumIndex() == 0 || currentCoordinate.getColumIndex()==numOfColumns-1){
				res.add(move);
			}else{
				if(isBlack(piece) && (!isWhite(boardDisplay[i][currentCoordinate.getColumIndex()+1]) || !isWhite(boardDisplay[i][currentCoordinate.getColumIndex()-1])))
					res.add(move);
				if(isWhite(piece) && (!isBlack(boardDisplay[i][currentCoordinate.getColumIndex()+1]) || !isBlack(boardDisplay[i][currentCoordinate.getColumIndex()-1])))
					res.add(move);
			}
		}
		return res;
	}


	private static Collection<? extends Move> checkRightMoves(Piece piece, Coordinate currentCoordinate, Integer numOfColumns, Integer numOfRows, Piece[][] boardDisplay, List<Coordinate> keySquares) {
		List<Move> res = new ArrayList<Move>();
		for(int j = currentCoordinate.getColumIndex()+1; j<numOfColumns; j++){
			Move move = Move.of(currentCoordinate.getRowIndex(), currentCoordinate.getColumIndex(), currentCoordinate.getRowIndex(), j);
			if(!boardDisplay[currentCoordinate.getRowIndex()][j].equals(Piece.EMPTY))
				break;
			if(keySquares.contains(new Coordinate(currentCoordinate.getRowIndex(), j)) && !piece.equals(Piece.WHITE_KING))
				continue;
			if(currentCoordinate.getRowIndex()==0 || currentCoordinate.getRowIndex()==numOfRows-1){
				res.add(Move.of(currentCoordinate, new Coordinate(currentCoordinate.getRowIndex(), j)));
			}else{
				if(isBlack(piece) && (!isWhite(boardDisplay[currentCoordinate.getRowIndex()+1][j]) || !isWhite(boardDisplay[currentCoordinate.getRowIndex()-1][j])))
					res.add(move);
				if(isWhite(piece) && (!isBlack(boardDisplay[currentCoordinate.getRowIndex()+1][j]) || !isBlack(boardDisplay[currentCoordinate.getRowIndex()-1][j])))
					res.add(move);
				}
		}
		return res;
	}


	private static Collection<Move> checkLeftMoves(Piece piece, Coordinate currentCoordinate, Integer numOfColumns, Integer numOfRows,Piece[][] boardDisplay, List<Coordinate> keySquares) {

		List<Move> res = new ArrayList<Move>();
		for(int j = currentCoordinate.getColumIndex()-1; j>=0; j--){
			Move move = Move.of(currentCoordinate.getRowIndex(), currentCoordinate.getColumIndex(), currentCoordinate.getRowIndex(), j);
			if(!boardDisplay[currentCoordinate.getRowIndex()][j].equals(Piece.EMPTY)){
				break;
			}
			if(keySquares.contains(new Coordinate(currentCoordinate.getRowIndex(), j)) && !piece.equals(Piece.WHITE_KING)){
				continue;
			}
			if(currentCoordinate.getRowIndex()==0 || currentCoordinate.getRowIndex()==numOfRows-1){
				res.add(Move.of(currentCoordinate, new Coordinate(currentCoordinate.getRowIndex(), j)));
			}else{
				if(isBlack(piece) && (!isWhite(boardDisplay[currentCoordinate.getRowIndex()+1][j]) || !isWhite(boardDisplay[currentCoordinate.getRowIndex()-1][j])))
					res.add(move);
				if(isWhite(piece) && (!isBlack(boardDisplay[currentCoordinate.getRowIndex()+1][j]) || !isBlack(boardDisplay[currentCoordinate.getRowIndex()-1][j])))
					res.add(move);
			}
		}
		return res;
	}


	/**
	 * 
	 * @param move The move to be executed
	 * @return a copy of the current board but the boardDisplay now reflects the changes
	 */
	public Board executeMove(Move move) {
		Board nextBoard = this.clone();
		Integer row = move.getInitialCoordinate().getRowIndex();
		Integer colum = move.getInitialCoordinate().getColumIndex();
		
		Integer nextRow = move.getFinalCoordinate().getRowIndex();
		Integer nextColum = move.getFinalCoordinate().getColumIndex();
		
		Piece p = nextBoard.boardDisplay[row][colum];
		nextBoard.boardDisplay[row][colum] = Piece.EMPTY;
		nextBoard.boardDisplay[nextRow][nextColum] = p;
		//Changes to the availableMoves map needs to be implemented
		return nextBoard;
	}
	
	
	/**
	 * 
	 * @param c The position of the piece we want to know the color of
	 * @return the color of the player if the position is occupied, null otherwise
	 */
	public PlayerColor getPlayerColorOfPiece(Coordinate c) {
		Piece p = boardDisplay[c.getRowIndex()][c.getColumIndex()];
		switch(p) {
		case BLACK_PAWN:
			return PlayerColor.BLACK;
		case WHITE_PAWN:
			return PlayerColor.WHITE;
		case WHITE_KING:
			return PlayerColor.WHITE;
		default:
			return null;
		}
	}

	public Piece getPieceAtCoordinate(Coordinate c){
		return boardDisplay[c.getRowIndex()][c.getColumIndex()];
	}

	public static boolean isWhite(Piece p){
		return p.equals(Piece.WHITE_KING) || p.equals(Piece.WHITE_PAWN);
	}

	public static boolean isBlack(Piece p){
		return p.equals(Piece.BLACK_PAWN);
	}


	/**
	 * 
	 * @return True if the king is on the board, False otherwise
	 */
	public boolean isKingAlive(){

		boolean res = false;

		for(Piece[] row : boardDisplay){
			for(Piece piece : row){
				if(piece.equals(Piece.WHITE_KING)){
					res = true;
					break;
				}
					
			}
		}
		return res;
	}

	public void imprimeTablero(){
		String res = "";

		for(int i = 0; i<this.getNumOfRows(); i++){
			for(int j = 0; j<this.getNumOfColumns(); j++){
				Piece piece = boardDisplay[i][j];
				if(piece.equals(Piece.EMPTY))
					res+="·  ";
				else if(piece.equals(Piece.BLACK_PAWN))
					res+="B  ";
				else if(piece.equals(Piece.WHITE_PAWN))
					res+="w  ";
				else
					res+="k  ";
			}
			res += "\n\n";
		}

		System.out.println(res);
		System.out.println("---------------------------------------------------------------------------");
	}


	public static Board checkPiecesTaken(Board nextBoard) {

		Board nextBoardChecked = nextBoard.clone();


		for(int i = 0; i<nextBoard.getNumOfRows(); i++){
            for(int j = 1; j<nextBoard.getNumOfColumns()-1; j++){
                Piece leftPiece = nextBoard.getBoardDisplay()[i][j-1];
                Piece middlePiece = nextBoard.getBoardDisplay()[i][j];
                Piece rightPiece = nextBoard.getBoardDisplay()[i][j+1];

				if(middlePiece.equals(Piece.BLACK_PAWN) && (leftPiece.equals(Piece.WHITE_KING) || leftPiece.equals(Piece.WHITE_PAWN)) && (rightPiece.equals(Piece.WHITE_KING) || rightPiece.equals(Piece.WHITE_PAWN)))
					nextBoardChecked.getBoardDisplay()[i][j] = Piece.EMPTY;

				if(middlePiece.equals(Piece.WHITE_PAWN) && leftPiece.equals(Piece.BLACK_PAWN) && rightPiece.equals(Piece.BLACK_PAWN))
					nextBoardChecked.getBoardDisplay()[i][j] = Piece.EMPTY;

                if(middlePiece.equals(Piece.WHITE_KING)){
					Coordinate coordinate = new Coordinate(i, j);
					if(coordinate.equals(nextBoard.getThrone()) && nextBoard.getPieceAtCoordinate(new Coordinate(i-1, j)).equals(Piece.BLACK_PAWN) && nextBoard.getPieceAtCoordinate(new Coordinate(i+1, j)).equals(Piece.BLACK_PAWN) && nextBoard.getPieceAtCoordinate(new Coordinate(i, j-1)).equals(Piece.BLACK_PAWN) && nextBoard.getPieceAtCoordinate(new Coordinate(i, j+1)).equals(Piece.BLACK_PAWN))
						nextBoardChecked.getBoardDisplay()[i][j] = Piece.EMPTY;
					else if(nextBoard.isSurroundingsOfThrone(coordinate)){
						List<Piece> piecesAround = new ArrayList<Piece>();
						piecesAround.add(nextBoard.getPieceAtCoordinate(new Coordinate(i-1, j)));
						piecesAround.add(nextBoard.getPieceAtCoordinate(new Coordinate(i+1, j)));
						piecesAround.add(nextBoard.getPieceAtCoordinate(new Coordinate(i, j-1)));
						piecesAround.add(nextBoard.getPieceAtCoordinate(new Coordinate(i, j+1)));
						if(piecesAround.stream().filter(new Predicate<Piece>() {
							public boolean test(Piece t) {
								return t.equals(Piece.BLACK_PAWN);
							}
						}).collect(Collectors.counting()) >= 3){
							nextBoardChecked.getBoardDisplay()[i][j] = Piece.EMPTY;
						}
					}
					else if(leftPiece.equals(Piece.BLACK_PAWN) && rightPiece.equals(Piece.BLACK_PAWN))
						nextBoardChecked.getBoardDisplay()[i][j] = Piece.EMPTY;
				}
            }
        }
        for(int j = 0; j<nextBoard.getNumOfColumns(); j++){
            for(int i = 1; i<nextBoard.getNumOfRows()-1; i++){
                Piece topPiece = nextBoard.getBoardDisplay()[i-1][j];
                Piece middlePiece = nextBoard.getBoardDisplay()[i][j];
                Piece bottomPiece = nextBoard.getBoardDisplay()[i+1][j];

                if(middlePiece.equals(Piece.BLACK_PAWN) && (topPiece.equals(Piece.WHITE_KING) || topPiece.equals(Piece.WHITE_PAWN)) && (bottomPiece.equals(Piece.WHITE_KING) || bottomPiece.equals(Piece.WHITE_PAWN)))
					nextBoardChecked.getBoardDisplay()[i][j] = Piece.EMPTY;
				if(middlePiece.equals(Piece.WHITE_PAWN) && topPiece.equals(Piece.BLACK_PAWN) && bottomPiece.equals(Piece.BLACK_PAWN))
					nextBoardChecked.getBoardDisplay()[i][j] = Piece.EMPTY;
				if(middlePiece.equals(Piece.WHITE_KING)){
					Coordinate coordinate = new Coordinate(i, j);
					if(!coordinate.equals(nextBoard.getThrone()) && nextBoard.isSurroundingsOfThrone(coordinate) && topPiece.equals(Piece.BLACK_PAWN) && bottomPiece.equals(Piece.BLACK_PAWN))
						nextBoardChecked.getBoardDisplay()[i][j] = Piece.EMPTY;
				}
            }
        }

		return nextBoardChecked;
	}
}

File: MonteCarloTreeSearch/src/main/python/game/Board.py

from cmath import pi
from Piece import Piece
from Coordinate import Coordinate
from PlayerColor import PlayerColor
from Move import Move
from copy import copy


class Board:

    def __init__(self, boardDisplay):
        self.boardDisplay = boardDisplay

    def integer_display_to_board(integerDisplay):
        boardDisplay = []
        for i in range(0, len(integerDisplay)):
            row = []
            for j in range(0, len(integerDisplay[i])):
                piece = integerDisplay[i][j]
                row.append(Piece._value2member_map_.get(piece))
            boardDisplay.append(row)
        return Board(boardDisplay)

    def copy(self):
        newBoardDisplay = []
        for i in range(0, len(self.boardDisplay)):
            row = []
            for j in range(0, len(self.boardDisplay[i])):
                piece = self.boardDisplay[i][j]
                if piece == Piece.EMPTY:
                    row.append(Piece.EMPTY)
                elif piece == Piece.BLACK_PAWN:
                    row.append(Piece.BLACK_PAWN)
                elif piece == Piece.WHITE_PAWN:
                    row.append(Piece.WHITE_PAWN)
                elif piece == Piece.WHITE_KING:
                    row.append(Piece.WHITE_KING)
                else:
                    row.append(piece)
            newBoardDisplay.append(row)

        newBoard = Board(newBoardDisplay)
        
        return Board(newBoardDisplay)


    def get_num_of_rows(self):
        return len(self.boardDisplay)

    def get_num_of_columns(self):
        return len(self.boardDisplay[0])

    def get_throne(self):
        return Coordinate((self.get_num_of_rows()-1)/2, (self.get_num_of_rows()-1)/2)

    def is_surroundings_of_throne(self, c):

        isMiddleTop = c.rowIndex == ((self.get_num_of_rows()-1)/2)-1 and c.columnIndex == (self.get_num_of_columns()-1)/2
        isMiddleBottom = c.rowIndex == ((self.get_num_of_rows()-1)/2)+1 and c.columnIndex == (self.get_num_of_columns()-1)/2
        isMiddleLeft = c.rowIndex == ((self.get_num_of_rows()-1)/2) and c.columnIndex == ((self.get_num_of_columns()-1)/2)-1
        isMiddleRight = c.rowIndex == ((self.get_num_of_rows()-1)/2) and c.columnIndex == ((self.get_num_of_columns()-1)/2)+1

        return isMiddleTop or isMiddleBottom or isMiddleRight or isMiddleLeft

    def is_black(piece):
        return piece == Piece.BLACK_PAWN

    def is_white(piece):
        return piece == Piece.WHITE_PAWN or piece == Piece.WHITE_KING
    
    def check_left_moves(piece, currentCoordinate, columns, rows, boardDisplay, keySquares):
        res = []
        for j in range(int(currentCoordinate.columnIndex)-1, -1, -1):
            move = Move(currentCoordinate.rowIndex, currentCoordinate.columnIndex, currentCoordinate.rowIndex, j)
            if boardDisplay[int(currentCoordinate.rowIndex)][j] != Piece.EMPTY:
                break
            if (Coordinate(currentCoordinate.rowIndex, j) in keySquares) and (piece != Piece.WHITE_KING):
                continue
            if currentCoordinate.rowIndex == 0 or currentCoordinate.rowIndex == rows-1:
                res.append(move)
            else:
                pieceDown = boardDisplay[currentCoordinate.rowIndex+1][j]
                pieceUp = boardDisplay[currentCoordinate.rowIndex-1][j]
                if Board.is_black(piece) and (not Board.is_white(pieceUp) or not Board.is_white(pieceDown)):
                    res.append(move)
                if Board.is_white(piece) and (not Board.is_black(pieceUp) or not Board.is_black(pieceDown)):
                    res.append(move)
        return res
    

    def check_right_moves(piece, currentCoordinate, columns, rows, boardDisplay, keySquares):
        res = []
        for j in range(int(currentCoordinate.columnIndex)+1, columns):
            move = Move(currentCoordinate.rowIndex, currentCoordinate.columnIndex, currentCoordinate.rowIndex, j)
            if boardDisplay[int(currentCoordinate.rowIndex)][j] != Piece.EMPTY:
                break
            if (Coordinate(currentCoordinate.rowIndex, j) in keySquares) and (piece != Piece.WHITE_KING):
                continue
            if currentCoordinate.rowIndex == 0 or currentCoordinate.rowIndex == rows-1:
                res.append(move)
            else:
                pieceDown = boardDisplay[currentCoordinate.rowIndex+1][j]
                pieceUp = boardDisplay[currentCoordinate.rowIndex-1][j]
                if Board.is_black(piece) and (not Board.is_white(pieceUp) or not Board.is_white(pieceDown)):
                    res.append(move)
                if Board.is_white(piece) and (not Board.is_black(pieceUp) or not Board.is_black(pieceDown)):
                    res.append(move)
        return res

    def check_up_moves(piece, currentCoordinate, columns, rows, boardDisplay, keySquares):
        res = []
        for i in range(int(currentCoordinate.rowIndex)-1, -1, -1):
            move = Move(currentCoordinate.rowIndex, currentCoordinate.columnIndex, i, currentCoordinate.columnIndex)
            if boardDisplay[i][int(currentCoordinate.columnIndex)] != Piece.EMPTY:
                break
            if (Coordinate(i, currentCoordinate.columnIndex) in keySquares) and (piece != Piece.WHITE_KING):
                continue
            if currentCoordinate.columnIndex == 0 or currentCoordinate.columnIndex == columns-1:
                res.append(move)
            else:
                pieceRight = boardDisplay[i][currentCoordinate.columnIndex+1]
                pieceLeft = boardDisplay[i][currentCoordinate.columnIndex-1]
                if Board.is_black(piece) and (not Board.is_white(pieceLeft) or not Board.is_white(pieceRight)):
                    res.append(move)
                if Board.is_white(piece) and (not Board.is_black(pieceLeft) or not Board.is_black(pieceRight)):
                    res.append(move)
        return res

    def check_down_moves(piece, currentCoordinate, columns, rows, boardDisplay, keySquares):
        res = []
        for i in range(int(currentCoordinate.rowIndex)+1, rows):
            move = Move(currentCoordinate.rowIndex, currentCoordinate.columnIndex, i, currentCoordinate.columnIndex)
            if boardDisplay[i][int(currentCoordinate.columnIndex)] != Piece.EMPTY:
                break
            if (Coordinate(i, currentCoordinate.columnIndex) in keySquares) and (piece != Piece.WHITE_KING):
                continue
            if currentCoordinate.columnIndex == 0 or currentCoordinate.columnIndex == columns-1:
                res.append(move)
            else:
                pieceRight = boardDisplay[i][currentCoordinate.columnIndex+1]
                pieceLeft = boardDisplay[i][currentCoordinate.columnIndex-1]
                if Board.is_black(piece) and (not Board.is_white(pieceLeft) or not Board.is_white(pieceRight)):
                    res.append(move)
                if Board.is_white(piece) and (not Board.is_black(pieceLeft) or not Board.is_black(pieceRight)):
                    res.append(move)
        return res

    

    def get_player_available_moves(self):
        res = dict()
        rows = self.get_num_of_rows()
        columns = self.get_num_of_columns()

        for i in range(0,rows):
            for j in range(0, columns):
                piece = self.boardDisplay[i][j]
                moves = Board.get_available_moves_of_piece(Coordinate(i,j), self.boardDisplay)

                if piece == Piece.BLACK_PAWN:
                    if PlayerColor.BLACK in res:
                        moveList = list(res[PlayerColor.BLACK])
                        moveList.extend(moves)
                        res[PlayerColor.BLACK] = moveList
                    else:
                        res[PlayerColor.BLACK] = moves
                elif piece == Piece.WHITE_PAWN or piece == Piece.WHITE_KING:
                    if PlayerColor.WHITE in res:
                        moveList = list(res[PlayerColor.WHITE])
                        moveList.extend(moves)
                        res[PlayerColor.WHITE] = moveList
                    else:
                        res[PlayerColor.WHITE] = moves
        return res

    def get_available_moves_of_piece(currentCoordinate, boardDisplay):
        res = []
        rows = len(boardDisplay)
        columns = len(boardDisplay[0])
        piece = boardDisplay[currentCoordinate.rowIndex][currentCoordinate.columnIndex]

        keySquares = [Coordinate(0,0), Coordinate(rows-1, 0), Coordinate(0, columns-1), Coordinate(rows-1, columns-1)]

        if piece == Piece.EMPTY:
            return res

        res.extend(Board.check_left_moves(piece, currentCoordinate, columns, rows, boardDisplay, keySquares))
        res.extend(Board.check_right_moves(piece, currentCoordinate, columns, rows, boardDisplay, keySquares))
        res.extend(Board.check_up_moves(piece, currentCoordinate, columns, rows, boardDisplay, keySquares))
        res.extend(Board.check_down_moves(piece, currentCoordinate, columns, rows, boardDisplay, keySquares))

        return res

    def get_piece_at_coordinate(self, c):
        return self.boardDisplay[c.rowIndex][c.columnIndex]

    def is_king_alive(self):
        res = False
        for row in self.boardDisplay:
            for piece in row:
                if piece == Piece.WHITE_KING:
                    return True
        return res

    def print_board(self):
        res = ""
        for i in range(0, self.get_num_of_rows()):
            for j in range(0, self.get_num_of_columns()):
                piece = self.boardDisplay[i][j]
                if piece == Piece.EMPTY:
                    res+="·  "
                elif piece == Piece.BLACK_PAWN:
                    res+="B  "
                elif piece == Piece.WHITE_PAWN:
                    res+="w  "
                else:
                    res+="k  "
            res += "\n\n"
        print(res)

    def check_pieces_taken(nextBoard):
        nextBoardChecked = nextBoard.copy()

        #Checkeo horizontal
        for i in range(0, nextBoard.get_num_of_rows()):
            for j in range(1, nextBoard.get_num_of_columns()-1):
                leftPiece = nextBoard.boardDisplay[i][j-1]
                middlePiece = nextBoard.boardDisplay[i][j]
                rightPiece = nextBoard.boardDisplay[i][j+1]

                if Board.is_black(middlePiece) and Board.is_white(leftPiece) and Board.is_white(rightPiece):
                    nextBoardChecked.boardDisplay[i][j] = Piece.EMPTY
                if middlePiece == Piece.WHITE_PAWN and Board.is_black(leftPiece) and Board.is_black(rightPiece):
                    nextBoardChecked.boardDisplay[i][j] = Piece.EMPTY
                
                if middlePiece == Piece.WHITE_KING:
                    coordinate = Coordinate(i,j)
                    if coordinate == nextBoard.get_throne() and nextBoard.get_piece_at_coordinate(Coordinate(i-1, j)) == Piece.BLACK_PAWN and nextBoard.get_piece_at_coordinate(Coordinate(i+1, j)) == Piece.BLACK_PAWN and nextBoard.get_piece_at_coordinate(Coordinate(i, j-1)) == Piece.BLACK_PAWN and nextBoard.get_piece_at_coordinate(Coordinate(i, j+1)) == Piece.BLACK_PAWN:
                        nextBoardChecked.boardDisplay[i][j] = Piece.EMPTY
                    elif nextBoard.is_surroundings_of_throne(coordinate):
                        piecesAround = []
                        piecesAround.append(nextBoard.get_piece_at_coordinate(Coordinate(i-1, j)))
                        piecesAround.append(nextBoard.get_piece_at_coordinate(Coordinate(i+1, j)))
                        piecesAround.append(nextBoard.get_piece_at_coordinate(Coordinate(i, j-1)))
                        piecesAround.append(nextBoard.get_piece_at_coordinate(Coordinate(i, j+1)))
                        blackPawnsCounter = 0
                        for pieceAround in piecesAround:
                            if pieceAround == Piece.BLACK_PAWN:
                                blackPawnsCounter += 1
                        if blackPawnsCounter >= 3:
                            nextBoardChecked.boardDisplay[i][j] = Piece.EMPTY
                    elif Board.is_black(rightPiece) and Board.is_black(leftPiece):
                        nextBoardChecked.boardDisplay[i][j] = Piece.EMPTY

        #Checkeo vertical

        for i in range(1, nextBoard.get_num_of_rows()-1):
            for j in range(0, nextBoard.get_num_of_columns()):
                topPiece = nextBoard.boardDisplay[i-1][j]
                middlePiece = nextBoard.boardDisplay[i][j]
                bottomPiece = nextBoard.boardDisplay[i+1][j]

                if Board.is_black(middlePiece) and Board.is_white(topPiece) and Board.is_white(bottomPiece):
                    nextBoardChecked.boardDisplay[i][j] = Piece.EMPTY
                if(middlePiece == Piece.WHITE_PAWN and Board.is_black(topPiece) and Board.is_black(bottomPiece)):
                    nextBoardChecked.boardDisplay[i][j] = Piece.EMPTY
                if(middlePiece == Piece.WHITE_KING and Coordinate(i,j) != nextBoard.get_throne() and not nextBoard.is_surroundings_of_throne(Coordinate(i,j)) and Board.is_black(topPiece) and Board.is_black(bottomPiece)):
                    nextBoardChecked.boardDisplay[i][j] = Piece.EMPTY
        return nextBoardChecked
                    
    
'''
integerDisplay = [[0,3,0,1,0],[0,2,2,0,1],[1,0,2,1,2],[2,0,0,0,0],[0,0,1,0,0]]
board = Board.integer_display_to_board(integerDisplay)


board.print_board()

moves = board.get_player_available_moves()[PlayerColor.WHITE]
for move in moves:
    print(move)
board2 = Board.check_pieces_taken(board)
board.print_board()
board2.print_board()
'''